Spigot-Server-Patches/0419-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,745 @@
+         final ObjectLinkedOpenHashSet<EntityPlayer> gone = new ObjectLinkedOpenHashSet<>(this.players.keySet());
+ 
+-        final int oldViewDistance = this.viewDistance;
+-        this.viewDistance = newViewDistance;
++        //Paper start ticket level pp-view-dist
++        //final int oldViewDistance = this.viewDistance;
++        //this.viewDistance = newViewDistance;
+ 
+         for (final EntityPlayer player : currentPlayers) {
++            player.oldMobViewDistance = player.mobViewDistance;
++            player.mobViewDistance = player.getTrackViewDistance();
++            //Paper end
++
+             if (player.isSpectator() || !player.affectsSpawning) {
+                 continue; // will be left in 'gone' (or not added at all)
+             }
+@@ -50,9 +55,10 @@ public final class PlayerMobDistanceMap {
+             final SectionPosition oldPosition = this.players.put(player, newPosition);
+ 
+             if (oldPosition == null) {
+-                this.addNewPlayer(player, newPosition, newViewDistance);
++                this.addNewPlayer(player, newPosition, player.mobViewDistance); //Paper ticket level pp-view-dist
+             } else {
+-                this.updatePlayer(player, oldPosition, newPosition, oldViewDistance, newViewDistance);
++                this.updatePlayer(player, oldPosition, newPosition, player.oldMobViewDistance, player.mobViewDistance); //Paper ticket level pp-view-dist
++
+             }
+             //this.validatePlayer(player, newViewDistance); // debug only
+         }
+@@ -60,7 +66,7 @@ public final class PlayerMobDistanceMap {
+         for (final EntityPlayer player : gone) {
+             final SectionPosition oldPosition = this.players.remove(player);
+             if (oldPosition != null) {
+-                this.removePlayer(player, oldPosition, oldViewDistance);
++                this.removePlayer(player, oldPosition, player.oldMobViewDistance); //Paper ticket level pp-view-dist
+             }
+         }
+     }
+diff --git a/src/main/java/net/minecraft/server/ChunkMapDistance.java b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+index 63a68872..4bf707da 100644
+--- a/src/main/java/net/minecraft/server/ChunkMapDistance.java
++++ b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+@@ -43,6 +43,7 @@ public abstract class ChunkMapDistance {
+     private final LongSet l = new LongOpenHashSet();
+     private final Executor m;
+     private long currentTick;
++    private ChunkMapDistance.c getPlayerTicketTracker() {return this.g;} //Paper OBFHELPER
+ 
+     protected ChunkMapDistance(Executor executor, Executor executor1) {
+         executor1.getClass();
+@@ -234,6 +235,7 @@ public abstract class ChunkMapDistance {
+         ((ObjectSet) this.c.computeIfAbsent(i, (j) -> {
+             return new ObjectOpenHashSet();
+         })).add(entityplayer);
++        this.g.update(entityplayer, i); //Paper ticket level pp-view-dist
```

Comment:
defining an obfhelper is no good if you do not use it

Reply:
fixed and will be committed